### ISSUE TITLE
Call for Contributions

### ISSUE THREAD

**Issue by houseroad** [onnx/onnx#426]
# Call for Contributions
ONNX is open source, and contributions from the community is welcome. Many tasks, (some are low-hanging fruits), are listed here. If you would like to contribute to the project, but don't know where to start, this list is worth to check.

## Adding ONNX Backend Test Case
Ideally, we should have sample cases for every operator, which can generate [backend test data](https://github.com/onnx/onnx/tree/master/onnx/backend/test/data/) and also serve as [examples](https://github.com/onnx/onnx/blob/master/docs/Operators.md#examples-1). However, many operators have no sample cases, and even not covered by ONNX backend test cases. So please contribute your useful cases to ONNX repository.

[How to contribute ONNX backend test?](https://github.com/onnx/onnx/blob/master/docs/OnnxBackendTest.md)

The operators which don't have examples:
- [x] Abs
- [x] Add
- [x] And
- [x] ArgMax
- [x] ArgMin
- [x] AveragePool
- [x] BatchNormalization
- [x] Cast
- [x] Ceil
- [x] Clip
- [x] Concat
- [x] Conv
- [x] ConvTranspose
- [x] DepthToSpace
- [x] Div
- [x] Dropout
- [x] Elu
- [x] Equal
- [x] Exp
- [x] Flatten
- [x] Floor
- [x] GRU
- [x] Gather
- [x] Gemm
- [x] GlobalAveragePool
- [ ] GlobalLpPool
- [x] GlobalMaxPool
- [x] Greater
- [x] HardSigmoid
- [x] Hardmax
- [x] InstanceNormalization
- [x] LRN
- [x] LSTM
- [x] LeakyRelu
- [x] Less
- [x] Log
- [x] LogSoftmax
- [x] LpNormalization
- [ ] LpPool
- [x] Max
- [x] MaxPool
- [ ] MaxRoiPool
- [x] Mean
- [x] Min
- [x] Mul
- [x] Neg
- [x] Not
- [x] Or
- [x] PRelu
- [x] Pow
- [x] RNN
- [x] RandomNormal
- [x] RandomNormalLike
- [x] RandomUniform
- [x] RandomUniformLike
- [x] Reciprocal
- [x] ReduceL1
- [x] ReduceL2
- [x] ReduceLogSum
- [x] ReduceLogSumExp
- [x] ReduceMax
- [x] ReduceMean
- [x] ReduceMin
- [x] ReduceProd
- [x] ReduceSum
- [x] ReduceSumSquare
- [x] Reshape
- [x] Selu
- [x] Sigmoid
- [x] Softmax
- [x] Softplus
- [x] Softsign
- [ ] SpaceToDepth
- [x] Split
- [x] Sqrt
- [x] Squeeze
- [x] Sub
- [x] Sum
- [x] Tanh
- [x] Tile
- [x] TopK
- [x] Transpose
- [x] Unsqueeze
- [x] Xor
- [ ] experimental ATen
- [ ] experimental Affine
- [ ] experimental ConstantFill
- [ ] experimental Crop
- [ ] experimental FC
- [ ] experimental GRUUnit
- [ ] experimental GivenTensorFill
- [ ] experimental Identity
- [ ] experimental If
- [ ] experimental ImageScaler
- [ ] experimental Loop
- [ ] experimental LoopIndexTensor
- [ ] experimental MeanVarianceNormalization
- [ ] experimental ParametricSoftplus
- [ ] experimental Scale
- [ ] experimental ScaledTanh
- [x] experimental ThresholdedRelu
- [x] experimental Upsample

## Adding Export Support in PyTorch
Some simple operators are not supported by PyTorch exporter yet, call for contributions.

[How to add ONNX support in PyTorch](https://github.com/onnx/tutorials/blob/master/tutorials/PytorchAddExportSupport.md)

[How to generate ONNX backend test data in PyTorch operator export test](https://github.com/onnxbot/onnx-fb-universe#generate-test-from-pytorch-operator)

Here is a TO-DO list:
- [x] Tile (Repeat in PyTorch)

**Comment by houseroad** [comment]
@guoyuhong, thanks for contributing the node test of `and`, `or`, `xor`, and `not`!

**Comment by linkerzhang** [comment]
To avoid duplicate work, ReduceMax, ReduceMin, ReduceSum, Conv, Flatten, and generator OPs are being taken care of by some folks now.

**Comment by tianleiwu** [comment]
I am adding tests for some logical and math operators:
 - Equal, Greater, Less 
 - Ceil, Div, Floor, LeakyRelu, Mul, Neg, Reciprocal, Selu, Sqrt, Sub

**Comment by kevin645** [comment]
Working on ArgMax, and ArgMin. Will try to complete the list as I understand the source.

I was reading through the source of the backend stub that is test_backend_test.py, It's only checking the proto format. How can I setup the full ONNX reference implementation for the Ops? Do I need to set it up with a supported backend? (caffe2/pytorch)

**Comment by houseroad** [comment]
@kevin645 Thanks a lot! I marked ArgMax and ArgMin as you picked them up. After you finish the case test, you can run [cmd_tools.py](https://github.com/onnx/onnx/blob/master/onnx/backend/test/cmd_tools.py) to generate the data for backend test. The test data will generated in [onnx/backend/test/data](https://github.com/onnx/onnx/tree/master/onnx/backend/test/data) folder. test_backend_test.py only runs the checker on the test data. To have a full reference execution on the generated data, you need to install a backend, such as caffe2 with onnx-caffe2, or cntk. (Right now, PyTorch can only export the model, there is no importer yet.)

**Comment by linkerzhang** [comment]
Cast Reshape Concat Split Transpose will be covered, @houseroad please mark them to avoid duplication.

**Comment by bddppq** [comment]
Gather added in https://github.com/onnx/onnx/pull/537 thanks to @huitseeker

**Comment by terrytangyuan** [comment]
@houseroad I am putting together an R package for ONNX [here](https://github.com/terrytangyuan/onnx-r). Would you like the ownership get transferred to ONNX Github organization? It would help with the exposure of the package later on.

**Comment by anirudhacharya** [comment]
Has the tests for the following operators - ReduceMax, ReduceMin, ReduceSum, Conv, Flatten,  been merged? If no one is doing ReduceMean and ReduceProd, I can take it up. Please confirm.

**Comment by houseroad** [comment]
As far as I know, no one is working on ReduceMax, ReduceMin, ReduceSum, ReduceMean, ReduceProd yet. Feel free to take them. @anirudhacharya :-)

**Comment by anirudhacharya** [comment]
@houseroad This comment - https://github.com/onnx/onnx/issues/426#issuecomment-359139085 says that someone is working on ReduceMax, ReduceMin and ReduceSum. And the checklist in the description section of this Issue has these three operators marked as done.

I will work on the ReduceProd and ReduceMean operator for now.

**Comment by houseroad** [comment]
@anirudhacharya sounds good to me. Thanks!

**Comment by anirudhacharya** [comment]
@onnx/facebook-onnx-contributors Can we revisit this list and verify if it is up to date? I have made a few updates but it might still be a bit out of sync. 
Also when do we tick off an operator in the list - 
 - When someone starts working on it or when the PR gets merged?
 - If a particular operator has tests in pytorch-converted/pytorch-operator but does not have node tests do we consider it as full test coverage?

**Comment by houseroad** [comment]
A1: When someone starts working on it, we will mark the op

A2: No, we still need node test, which serves as doc purpose.

**Comment by anirudhacharya** [comment]
Then what is the purpose of pytorch-converted/pytorch-operator tests?

**Comment by anirudhacharya** [comment]
I am working on BatchNormalization, GEMM, and LRN operator tests.

**Comment by prasanthpul** [comment]
@houseroad  Suggest we close this old issue and open a new one if needed with what remains

**Comment by houseroad** [comment]
@prasanthpul okay, close it
